File: 0138_copyListWithRandomPointer/main.cpp
#include <unordered_map>
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = nullptr;
        random = nullptr;
    }
};


class Solution {
public:
    Node* copyRandomList(Node* head) {
        std::unordered_map<Node*, Node*> map;
        if(head == nullptr)
            return nullptr;
        if(map.find(head) != map.end())
            return map[head];
        
        Node *newNode = new Node(head->val);
        map[head] = newNode;
        newNode->next = copyRandomList(head->next);
        newNode->next = copyRandomList(head->random);
        return newNode;
    }
};

class Solution1 {
public:
    Node* copyRandomList(Node* head) {
        if(!head) 
            return nullptr;
            
        std::unordered_map<Node*, Node*> map;
        map[nullptr] = nullptr;
        Node *ptr = head;
        Node *prev = nullptr;
        while(ptr) {
            if(map.find(ptr) == map.end()) 
                map[ptr] = new Node(ptr->val);

            if(map.find(ptr->random) == map.end())
                map[ptr->random] = new Node(ptr->random->val);
            map[ptr]->random = map[ptr->random];

            if(prev) 
                prev->next = map[ptr];
            prev = map[ptr];
            ptr = ptr->next;
        }

        return map[head];
    }
};

class Solution2 {
public:
    Node* copyRandomList(Node* head) {
        if(!head) 
            return nullptr;
            
        std::unordered_map<Node*, Node*> map;
        map[nullptr] = nullptr;

        Node *ptr = head;
        Node *newNode;
        while(ptr) {
            newNode = new Node(ptr->val);
            map[ptr] = newNode;
            ptr = ptr->next;
        }

        ptr = head;
        while(ptr) {
            newNode = map[ptr];
            newNode->next = map[ptr->next];
            newNode->random = map[ptr->random];
            ptr = ptr->next;
        }
        
        return map[head];
    }
};

File: 0138_copyListWithRandomPointer/main.py
from typing import Optional

# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        hashMap = {}
        # hashMap = {None : None}

        cur = head
        while cur:
            newNode = Node(cur.val)
            hashMap[cur] = newNode
            cur = cur.next
        
        cur = head
        while cur:
            newNode = hashMap[cur]
            newNode.next = hashMap.get(cur.next, None)
            newNode.random = hashMap.get(cur.random, None)
            cur = cur.next
        
        return hashMap[head]
    
class Solution1:
    def copyRandomList(self, head: 'Node') -> 'Node':
        hashMap = {}
        def deep_copy(node):
            if not node: 
                return None
            if node in hashMap: 
                return hashMap[node]
            
            newNode = Node(node.val)
            hashMap[node] = newNode
            newNode.next = deep_copy(node.next)
            newNode.random = deep_copy(node.random)
            return newNode
        return deep_copy(head)
